### ISSUE TITLE
Is it possible to let user select mfa preference in login when both TOTP and SMS are enabled ? I did not see setUserMfaPreference in amplify Auth class. Is there any way to enable both totp and sms but not give preference to any (let user select it in login) ?

### ISSUE THREAD

**Issue by akromswn** [aws-amplify/amplify-js#4646]
** Which Category is your question related to? **
Auth
** What AWS Services are you utilizing? **
 Cognito
** Provide additional details e.g. code snippets **
aws-amplify version: ^1.1.38

**Comment by jordanranz** [comment]
Hey @akromswn,

This is possible using the utility component: [`SelectMfaType`](https://aws-amplify.github.io/docs/js/authentication#letting-user-select-mfa-type).

Under the hood this uses: [`Auth.setPreferredMFA`](https://aws-amplify.github.io/docs/js/authentication#setup-mfa-type), if you would like to continue just using the Auth class with custom UI

**Comment by akromswn** [comment]
@jordanranz Thanks for the response. But it seems you did not understand my question. I am already using Auth.setPreferredMFA to set preferred mfa method but it only has three options ( "TOTP" | "SMS" | "NOMFA" ).  But what I want is that I want to enable both `TOTP` and `SMS` but I do not want to set preference so that user can select it while he is logging in.

**Comment by akromswn** [comment]
<img width="568" alt="Screenshot 2019-12-29 at 5 52 15 PM" src="https://user-images.githubusercontent.com/54656843/71554600-8da94b80-2a64-11ea-8d6c-40ca7ce7f3bd.png">

**Comment by jordanranz** [comment]
Ok, apologies for misunderstanding. To clarify, you would like a single user to have the option to perform MFA with the option of both TOTP and SMS?

**Comment by akromswn** [comment]
@jordanranz yes that's right, at the moment I can set preference to `TOTP` or `SMS` while user is logged and he will be verified depending on that preference next time. So I would like the user to be able to login using any of both methods if he has set up both MFA methods. 

I was told that it is possible to do that using [cognitoUser.setUserMfaPreference](https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_SetUserMFAPreference.html) by enabling both methods but setting preference to `false`.

Sample code in [amazon-cognito-identity-js](https://www.npmjs.com/package/amazon-cognito-identity-js)

```javascript
const smsMfaSettings = {
    PreferredMfa: false,
    Enabled: true,
};
const totpMfaSettings = {
    PreferredMfa: false,
    Enabled: true,
};
cognitoUser.setUserMfaPreference(smsMfaSettings, totpMfaSettings, function(err, result) {
    if (err) {
        alert(err.message || JSON.stringify(err));
    }
    console.log('call result ' + result);
});
```
I think amplify uses amazon-cognito-identity-js at its core so I was wondering if it is possible to do something similar.

**Comment by jordanranz** [comment]
If it is possible with `amazon-cognito-identity-js` then it is possible with Amplify. This specific use case may need a code change though. Here is the code in question:
https://github.com/aws-amplify/amplify-js/blob/510f86ef39bae3fba9a17deb630237a64c9dc7dc/packages/auth/src/Auth.ts#L694-L784

It seems to only allow a single MFA type within the function. Maybe we can override the function to allow for multiple MFA types to be set.

I'll mark this as a feature request.

**Comment by dautushenka** [comment]
Any updates here?

**Comment by darrenfurr** [comment]
@jordanranz - Using setPreferredMfa during signIn does not work because the userSession is null.

So you get: `User is not authenticated`

Which occurs here:
```
function getUserData(callback, params) {
    if (!(this.signInUserSession != null && this.signInUserSession.isValid())) {
      this.clearCachedUserData();
      return callback(new Error('User is not authenticated'), null);
    }
```

I think the workaround is to use a lambda to make an adminSetUserMFAPreference call. 
Unfortunately though at this point in the auth process you do not even have a JWT token, so I'm not sure how you secure that endpoint.

**Comment by SaffanBaig** [comment]
Any update on this feature request? Currently I can set only one type of preferred MFA at a time for a user. Is there any other possible way to have this feature?

**Comment by nmlynch94** [comment]
Would also be curious to hear any workarounds people may have discovered in the meantime.

**Comment by darrenfurr** [comment]
We currently have our Cognito Pool set to MFA: Optional & then in the frontend, we allow the user to select their MFA option...
![Screen Shot 2022-03-31 at 2 37 49 PM](https://user-images.githubusercontent.com/133894/162027021-ed7cac89-a466-4e04-a3d3-0508599841f7.png)

```
 const user = await Auth.signIn(email.toLowerCase(), password);
      if (user.preferredMFA === 'NOMFA') {
       // load MFA Setup screen (above)
      }
```

Then we use `await Auth.setPreferredMFA(user, userSelectedOption);`

**Comment by helgabalashova** [comment]
Hi @akromswn 
It is possible to let user choose what MFA to use when they log in. But there are certain requirements for that to work:

1. both MFA methods should be set up for user
2. preferred MFA method should **NOT** be set up

If both of those requirements are in place, then every time user is trying to log in, it will get a `SELECT_MFA_TYPE` challenge from Cognito. In this case, you will be able to display two options and let user proceed with whatever they choose. 
Just remember, that you don't call `setPreferredMFA` method at any point. Once you call that function, you won't be able to go back to choosing MFA method during log in. 

The easiest way to set it up:

1. Set MFA as required
2. Enable both SMS and TOTP
3. Require user to enter phone number and confirm it at registration (it will enable SMS MFA by default for the user but will not set it as preferred type)
4. Ask user to set up TOTP after first login (first login will have SMS MFA) using `Auth.setUpTOTP` and `Auth.verifyTOTPToken` (**WITHOUT** calling `Auth.setPreferredMFA`)

next time user is trying to log in, they will get `SELECT_MFA_TYPE` challenge and either SMS or TOTP should work in this case.
Please, let us know if this solution works for you.
